rust/kernel/fs.rs

```diff
@@ -0,0 +1,1262 @@
+            drop_inode: None,
+            evict_inode: None,
+            put_super: None,
+            sync_fs: None,
+            freeze_super: None,
+            freeze_fs: None,
+            thaw_super: None,
+            unfreeze_fs: None,
+            statfs: Some(Self::statfs_callback),
+            remount_fs: None,
+            umount_begin: None,
+            show_options: None,
+            show_devname: None,
+            show_path: None,
+            show_stats: None,
+            #[cfg(CONFIG_QUOTA)]
+            quota_read: None,
+            #[cfg(CONFIG_QUOTA)]
+            quota_write: None,
+            #[cfg(CONFIG_QUOTA)]
+            get_dquots: None,
+            nr_cached_objects: None,
+            free_cached_objects: None,
+            shutdown: None,
+        };
+
+        unsafe extern "C" fn alloc_inode_callback(
+            sb: *mut bindings::super_block,
+        ) -> *mut bindings::inode {
+            // SAFETY: The callback contract guarantees that `sb` is valid for read.
+            let super_type = unsafe { (*sb).s_type };
+
+            // SAFETY: This callback is only used in `Registration`, so `super_type` is necessarily
+            // embedded in a `Registration`, which is guaranteed to be valid because it has a
+            // superblock associated to it.
+            let reg = unsafe { &*container_of!(super_type, Registration, fs) };
+
+            // SAFETY: `sb` and `cache` are guaranteed to be valid by the callback contract and by
+            // the existence of a superblock respectively.
+            let ptr = unsafe {
+                bindings::alloc_inode_sb(sb, MemCache::ptr(&reg.inode_cache), bindings::GFP_KERNEL)
+            }
+            .cast::<INodeWithData<T::INodeData>>();
+            if ptr.is_null() {
+                return ptr::null_mut();
+            }
+            ptr::addr_of_mut!((*ptr).inode)
+        }
+
+        unsafe extern "C" fn destroy_inode_callback(inode: *mut bindings::inode) {
+            // SAFETY: By the C contrat, inode is a valid pointer.
+            let is_bad = unsafe { bindings::is_bad_inode(inode) };
+
+            // SAFETY: The inode is guaranteed to be valid by the callback contract. Additionally,
+            // the superblock is also guaranteed to still be valid by the inode existence.
+            let super_type = unsafe { (*(*inode).i_sb).s_type };
+
+            // SAFETY: This callback is only used in `Registration`, so `super_type` is necessarily
+            // embedded in a `Registration`, which is guaranteed to be valid because it has a
+            // superblock associated to it.
+            let reg = unsafe { &*container_of!(super_type, Registration, fs) };
+            let ptr = container_of!(inode, INodeWithData<T::INodeData>, inode).cast_mut();
+
+            if !is_bad {
+                // SAFETY: The code either initialises the data or marks the inode as bad, since
+                // it's not bad, it's safey to drop it here.
```

Comment:
```suggestion
                // SAFETY: The code either initialises the data or marks the inode as bad, since
                // it's not bad, it's safe to drop it here.
```

Or, what about...

```suggestion
                // SAFETY: The code either initialises the data or marks the inode as bad. Since
                // the inode is not bad, the data is initialised, and thus safe to drop here.
```